File: Objective-C/TCMPPDemo/TCMPPMenuView.m
//
//  TCMPPMenuView.m
//  TCMPPDemo
//
//  Created by gavinjwxu on 2024/8/1.
//

#import "TCMPPMenuView.h"
#import "UIView+TZLayout.h"
#import "UIView+TCMPP.h"
#import "UIColor+TCMPP.h"

@interface TCMPPMenuView ()

@property (strong,nonatomic) NSArray<TCMPPMenuItem *> *menuItems;

@end

@implementation TCMPPMenuView

- (instancetype)initWithMenuItems:(NSArray<TCMPPMenuItem *> *)menuItems {
    if (self  = [super initWithFrame:CGRectZero]) {
        _menuItems = menuItems;
        self.backgroundColor = [UIColor tcmpp_colorWithHex:@"#4C4C4C"];
        [menuItems enumerateObjectsUsingBlock:^(TCMPPMenuItem *obj, NSUInteger idx, BOOL * _Nonnull stop) {
            UIView *content = [[UIView alloc] initWithFrame:CGRectMake(0, 50 * idx, 130, 50)];
            content.tag = idx;
            UITapGestureRecognizer *tap = [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(handleTapGesture:)];
            [content addGestureRecognizer:tap];
            [self addSubview:content];
            
            UIImageView *iv = [[UIImageView alloc] initWithFrame:CGRectMake(13, 13, 24, 24)];
            iv.image = obj.icon;
            [content addSubview:iv];
            
            UILabel *title = [[UILabel alloc] initWithFrame:CGRectMake(iv.tz_right + 10, 13, 130 - 13 - 10 - 24, 24)];
            title.font = [UIFont systemFontOfSize:15];
            title.textColor = UIColor.whiteColor;
            title.text = obj.title;
            [content addSubview:title];
        }];
    }
    return self;
}

- (void)dealloc {
    
}

- (void)handleTapGesture:(UITapGestureRecognizer *)tap {
    self.menuItems[tap.view.tag].block();
    [self dismiss];
}

- (void)dismiss {
    [self removeFromSuperview];
}

@end

@implementation TCMPPMenuItem

@end

File: Swift/TCMPPDemo-Swift/Class/TCMPPMenuView.swift
//
//  TCMPPMenuView.swift
//  TCMPPDemo-Swift
//
//  Created by gavinjwxu on 2024/8/14.
//

import UIKit

class TCMPPMenuItem: NSObject {
    var title: String?
    var icon: UIImage?
    var block: (() -> Void)?
}

class TCMPPMenuView: UIView {
    var menuItems: [TCMPPMenuItem] = []
    
    init(menuItems: [TCMPPMenuItem]) {
        super.init(frame: .zero)
        self.menuItems = menuItems
        self.backgroundColor = UIColor.tcmpp_color(withHex: "#4C4C4C")
        for (idx, obj) in menuItems.enumerated() {
            let content = UIView(frame: CGRect(x: 0, y: 50 * idx, width: 130, height: 50))
            content.tag = idx
            let tap = UITapGestureRecognizer(target: self, action: #selector(handleTapGesture(_:)))
            content.addGestureRecognizer(tap)
            self.addSubview(content)
            
            let iv = UIImageView(frame: CGRect(x: 13, y: 13, width: 24, height: 24))
            iv.image = obj.icon
            content.addSubview(iv)
            
            let title = UILabel(frame: CGRect(x: iv.frame.maxX + 10, y: 13, width: 130 - 13 - 10 - 24, height: 24))
            title.font = UIFont.systemFont(ofSize: 15)
            title.textColor = .white
            title.text = obj.title
            content.addSubview(title)
        }
    }
    
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    @objc func handleTapGesture(_ tap: UITapGestureRecognizer) {
        self.menuItems[tap.view!.tag].block?()
        dismiss()
    }
    
    func dismiss() {
        removeFromSuperview()
    }
}
